### PR TITLE
Don't close settings dialogs invalidated on save and use on PVR Timer Settings

### PR DESCRIPTION
**Pull request by phunkyfish**
## Description

Proper solution which reverts https://github.com/xbmc/xbmc/pull/19050 which was a work around forcing a default value as it was the only possible solution. Thanks to @Montellese for the guidance on the correct way t fix this.

## Motivation and Context

Any PVR add-on could receive a invalid manual timer. Instead of fixing in each add-on it is better to fix in core.

## How Has This Been Tested?

On OSX creating various manual timers with pvr.vuplus

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: addons/resource.language.en_gb/resources/strings.po [hunk 1]
```diff
@@ -9597,7 +9597,11 @@ msgctxt "#19071"
 msgid "Update interval"
 msgstr ""
 
-#empty string with id 19072
+#. Dialog warning text for invalid timer setting
+#: xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp
+msgctxt "#19072"
+msgid "In order to add/update a timer the end date and time must be greater than the start date and time."
+msgstr ""
```
Comment (ksooo) [inline addons/resource.language.en_gb/resources/strings.po:9604]: This change belongs into the pvr commit of this PR.

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp [hunk 2]
```diff
@@ -145,8 +145,8 @@ void CGUIDialogLibExportSettings::OnSettingAction(const std::shared_ptr<const CS
 
   const std::string &settingId = setting->GetId();
 
-  if (settingId == CSettings::SETTING_MUSICLIBRARY_EXPORT_FOLDER && 
-      !m_settings.IsToLibFolders() && !m_settings.IsArtistFoldersOnly())
+  if (settingId == CSettings::SETTING_MUSICLIBRARY_EXPORT_FOLDER && !m_settings.IsToLibFolders() &&
```
Comment (Montellese) [inline xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp:148]: Looks unrelated.

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp [hunk 3]
```diff
@@ -296,8 +298,8 @@ void CGUIDialogLibExportSettings::UpdateToggles()
 }
 
 void CGUIDialogLibExportSettings::UpdateDescription()
-{ 
-  if (m_settings.IsToLibFolders())
```
Comment (Montellese) [inline xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp:300]: Looks unrelated.

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp [hunk 4]
```diff
@@ -381,7 +383,7 @@ void CGUIDialogLibExportSettings::InitializeSettings()
   }
   else
    items = m_settings.GetExportItems();
-
```
Comment (Montellese) [inline xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp:384]: Looks unrelated.

File: xbmc/settings/dialogs/GUIDialogSettingsBase.cpp [hunk 5]
```diff
@@ -193,9 +193,15 @@ bool CGUIDialogSettingsBase::OnMessage(CGUIMessage& message)
...
-        return true;
+        if (OnOkay())
+        {
+          Close();
+          return true;
+        }
+        else
+        {
+          return false;
+        }
```
Comment (Montellese) [inline xbmc/settings/dialogs/GUIDialogSettingsBase.cpp:204]: Maybe write
```cpp
if (!OnOkay())
  return false;

Close();
return true;
```

Comment (Montellese) [reply xbmc/settings/dialogs/GUIDialogSettingsBase.cpp:204]: Or to be consistent with the `CGUIDialogSettingsManagerBase::OnOkay()`:
```cpp
if (OnOkay())
{
  Close();
  return true;
}

return false;
```
Either way the style should be consistent :-)

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp [hunk 6]
```diff
@@ -589,7 +591,12 @@ bool CGUIDialogPVRTimerSettings::Save()
...
     {
-      CLog::Log(LOGWARNING, "Timer settings dialog: Specified recording end time < start time: expect errors!");
+      CLog::Log(LOGWARNING,
+                "Timer settings dialog: Specified end time must be greater than start time: "
+                "please fix before adding/updating a timer");
+      HELPERS::ShowOKDialogText(CVariant{19065}, // "Timer settings"
+                                CVariant{19072}); // In order to add/update a timer
+      bValid = false;
     }
     m_timerInfoTag->SetStartFromLocalTime(m_startLocalTime);
```
Comment (Montellese) [inline xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:601]: Is it intended that you already update `m_timerInfoTag` even though the input going into it has been considered invalid? Can't you just completely abort here and only execute the rest of the code if the input is valid?

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp [hunk 7]
```diff
@@ -589,7 +591,12 @@ bool CGUIDialogPVRTimerSettings::Save()
     }
     else if (m_endLocalTime < m_startLocalTime) // Assume the user knows what they are doing, but log a warning just in case
     {
-      CLog::Log(LOGWARNING, "Timer settings dialog: Specified recording end time < start time: expect errors!");
+      CLog::Log(LOGWARNING,
```
Comment (Montellese) [inline xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:594]: What use is it to log this here? It's a user error and the user is notified about it and has to fix it before (s)he can continue.

**Review by Montellese — COMMENTED** [review]
The changes to the settings dialogs look good apart from the minors I commented on.

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp (hunk 2 above)
Comment (ksooo) [reply xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp:148]: clang-format?

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp (hunk 7 above)
Comment (ksooo) [reply xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:594]: fair

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp (hunk 3 above)
Comment (ksooo) [reply xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp:300]: clang-format again

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp (hunk 2 above)
Comment (Montellese) [reply xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp:148]: I know but it touches code that is unrelated to this PR (same below).

Comment (phunkyfish) [reply xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp:148]: It may have auto trim on save in my IDE. Will check it later.

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp (hunk 6 above)
Comment (phunkyfish) [reply xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:601]: Good point.

**Comment by phunkyfish** [comment]
Thanks for the reviews. All issues should be addressed now.

**Review by ksooo — APPROVED** [review]

**Comment by phunkyfish** [comment]
@ksooo, I had to modify the PVR commit. Whereas it worked for adding a new timer it could partially modify a timer if updating where validation failed.

**Comment by emveepee** [comment]
I looked at the code and I suspect a hole in the logic so compiled the PR and it does  get it into the "can't be reached" logic when I try and modify from this screen This is an EPG based recording where the start/end dates used to be 0 as the start and end time but I moved that to a date in the past because of negative times.    

image

**Comment by phunkyfish** [comment]
Can you explain in a bit more detail how to reproduce this?

**Comment by emveepee** [comment]
It is a start/end any time (no spinners) EPG based recording.  The original time it was scheduled indicates a date rollover



Martin

**Comment by phunkyfish** [comment]
> 
> Sorry I can't be more help I was confused with this logic handling for EPG based recordings before when it was crashing trying to format a display tag that shouldn't be displayed and not it is now more complicated. I actually expected to get the prompt.

But your timer type does not show start and end time so it shouldn’t show the prompt.

**Comment by emveepee** [comment]
> 
> 
> > Sorry I can't be more help I was confused with this logic handling for EPG based recordings before when it was crashing trying to format a display tag that shouldn't be displayed and not it is now more complicated. I actually expected to get the prompt.
> 
> But your timer type does not show start and end time so it shouldn’t show the prompt.

Right no prompt but it doesn't make it to the addon and there is a message to check the logs but no error is logged.  Before the validate logic there is a log message that the channel uid is not found but this is any channel rule.

**Comment by phunkyfish** [comment]
Do you have the log message? Are you creating a new timer or updating an old one?

**Comment by emveepee** [comment]
Nothing is logged.  However you are right it isn't working in the nightly either there is different behavior in Matrix because it is working in Leia.  I notice now Leia handles this dialog differently.  It will probably have something to do with how Matrix is mapping channel based recordings to "All Channels" when the source channel uid is not found.

**Comment by phunkyfish** [comment]
> Edit. If I save it with a valid channel and then to any channel it works so you can probably ignore my comments as it is another Matrix issue

Ok. If you think it’s a core issue please create one and I can look at the weekend.

**Comment by emveepee** [comment]
> 
> Ok. If you think it’s a core issue please create one and I can look at the weekend.

I will put in a PR that does addresses the issue.  Whether it is an appropriate fix can be discussed.

**Comment by phunkyfish** [comment]
Sounds good.

**Review by Montellese — APPROVED** [review]
Looks good now.
